src/types.lisp

```diff
@@ -1034,24 +1034,64 @@ The buffer has dynamic extent and may be stack allocated."
 (defctype :llong  :long-long)
 (defctype :ullong :unsigned-long-long)
 
+(defmacro defctype-matching (name size-or-type base-types &key (match-by '=))
+  (let* ((target-size (typecase size-or-type
+                        (integer size-or-type)
+                        (T (foreign-type-size size-or-type))))
+         (matching-type (loop for type in base-types
+                              for size = (foreign-type-size type)
+                              do (when (funcall match-by target-size size)
+                                   (return type)))))
+    (if matching-type
+        `(defctype ,name ,matching-type)
+        `(warn #+sbcl 'sb-int:simple-compiler-note #-sbcl 'alexandria:simple-style-warning
```

Comment:
sb-int is an internal package isn't it? What's the rationale anyway?

Reply:
A style warning doesn't seem quite appropriate, and neither does a full warning. The compiler note seems like the closest fitting condition type.

Reply:
A full warning would likely result in a bug report which would be nice. In any case, if we could avoid using SBCL internals, that'd be nice. Don't want to anger Douglas or Stas. :D